qa_fecapi_repetition is flaky

Debian build logs show that qa_fecapi_repetition fails on i386:

https://buildd.debian.org/status/fetch.php?pkg=gnuradio&arch=i386&ver=3.10.9.0%7Erc1-2&stamp=1703091669&raw=0
https://buildd.debian.org/status/fetch.php?pkg=gnuradio&arch=i386&ver=3.10.9.0%7Erc1-3&stamp=1703129935&raw=0

```
        Start 118: qa_fecapi_repetition
118/264 Test #118: qa_fecapi_repetition .........................***Failed    1.54 sec
FFFF........
======================================================================
FAIL: test_async_00 (__main__.test_fecapi_repetition.test_async_00) (packed=True, rev_pack=True)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/gr-fec/python/fec/qa_fecapi_repetition.py", line 174, in test_async_00
    self.assertSequenceEqualGR(data_in, data_out)
  File "/<<PKGBUILDDIR>>/gnuradio-runtime/python/gnuradio/gr_unittest.py", line 115, in assertSequenceEqualGR
    self.assertEqual(len(data_in), len(data_out), msg="Lengths do not match")
AssertionError: 15360 != 14848 : Lengths do not match

======================================================================
FAIL: test_async_00 (__main__.test_fecapi_repetition.test_async_00) (packed=True, rev_pack=False)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/gr-fec/python/fec/qa_fecapi_repetition.py", line 174, in test_async_00
    self.assertSequenceEqualGR(data_in, data_out)
  File "/<<PKGBUILDDIR>>/gnuradio-runtime/python/gnuradio/gr_unittest.py", line 115, in assertSequenceEqualGR
    self.assertEqual(len(data_in), len(data_out), msg="Lengths do not match")
AssertionError: 15360 != 14848 : Lengths do not match

======================================================================
FAIL: test_async_00 (__main__.test_fecapi_repetition.test_async_00) (packed=False, rev_pack=True)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/gr-fec/python/fec/qa_fecapi_repetition.py", line 174, in test_async_00
    self.assertSequenceEqualGR(data_in, data_out)
  File "/<<PKGBUILDDIR>>/gnuradio-runtime/python/gnuradio/gr_unittest.py", line 115, in assertSequenceEqualGR
    self.assertEqual(len(data_in), len(data_out), msg="Lengths do not match")
AssertionError: 15360 != 15296 : Lengths do not match

======================================================================
FAIL: test_async_00 (__main__.test_fecapi_repetition.test_async_00) (packed=False, rev_pack=False)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/gr-fec/python/fec/qa_fecapi_repetition.py", line 174, in test_async_00
    self.assertSequenceEqualGR(data_in, data_out)
  File "/<<PKGBUILDDIR>>/gnuradio-runtime/python/gnuradio/gr_unittest.py", line 115, in assertSequenceEqualGR
    self.assertEqual(len(data_in), len(data_out), msg="Lengths do not match")
AssertionError: 15360 != 15296 : Lengths do not match

----------------------------------------------------------------------
Ran 9 tests in 1.202s

FAILED (failures=4)
```

--- a/gr-fec/lib/async_decoder_impl.cc
+++ b/gr-fec/lib/async_decoder_impl.cc
@@ -15,6 +15,7 @@
 #include "async_decoder_impl.h"
 #include <gnuradio/io_signature.h>
 #include <volk/volk.h>
+#include <cmath>
 #include <cstdio>
 
 namespace gr {
@@ -54,7 +55,7 @@ async_decoder_impl::async_decoder_impl(generic_decoder::sptr my_decoder,
     set_msg_handler(d_in_port, [this](const pmt::pmt_t& msg) { this->decode(msg); });
 
     // The maximum frame size is set by the initial frame size of the decoder.
-    d_max_bits_in = d_mtu * 8 * 1.0 / d_decoder->rate();
+    d_max_bits_in = std::lround(d_mtu * 8 * 1.0 / d_decoder->rate());
     d_tmp_f32.resize(d_max_bits_in);
 
     if (strncmp(d_decoder->get_input_conversion(), "uchar", 5) == 0) {
@@ -97,14 +98,14 @@ void async_decoder_impl::decode(const pmt::pmt_t& msg)
     // Watch out for this diff. It might be over-specializing to the
     // CC decoder in terminated mode that has an extra rate(K-1)
     // bits added on to the transmitted frame.
-    int diff =
-        d_decoder->rate() * d_decoder->get_input_size() - d_decoder->get_output_size();
+    int diff = std::lround(d_decoder->rate() * d_decoder->get_input_size()) -
+               d_decoder->get_output_size();
 
     size_t nbits_in = pmt::length(bits);
     size_t nbits_out = 0;
     size_t nblocks = 1;
     bool variable_frame_size =
-        d_decoder->set_frame_size(nbits_in * d_decoder->rate() - diff);
+        d_decoder->set_frame_size(std::lround(nbits_in * d_decoder->rate()) - diff);
 
     // Check here if the frame size is larger than what we've
     // allocated for in the constructor.
@@ -115,7 +116,7 @@ void async_decoder_impl::decode(const pmt::pmt_t& msg)
 
     // set up nbits_out
     if (variable_frame_size) {
-        nbits_out = nbits_in * d_decoder->rate() - diff;
+        nbits_out = std::lround(nbits_in * d_decoder->rate()) - diff;
     } else {
         nblocks = nbits_in / d_decoder->get_input_size();
         nbits_out = nblocks * d_decoder->get_output_size();

--- a/gr-fec/lib/async_encoder_impl.cc
+++ b/gr-fec/lib/async_encoder_impl.cc
@@ -16,6 +16,7 @@
 #include <gnuradio/io_signature.h>
 #include <gnuradio/logger.h>
 #include <volk/volk.h>
+#include <cmath>
 
 namespace gr {
 namespace fec {
@@ -55,7 +56,7 @@ async_encoder_impl::async_encoder_impl(generic_encoder::sptr my_encoder,
     if (d_packed) {
         set_msg_handler(d_in_port, [this](pmt::pmt_t msg) { this->encode_packed(msg); });
 
-        int max_bits_out = d_encoder->rate() * d_mtu * 8;
+        int max_bits_out = std::lround(d_encoder->rate() * d_mtu * 8);
         d_bits_out.resize(max_bits_out);
     } else {
         set_msg_handler(d_in_port,

--- a/gr-fec/lib/tagged_decoder_impl.cc
+++ b/gr-fec/lib/tagged_decoder_impl.cc
@@ -14,6 +14,7 @@
 
 #include "tagged_decoder_impl.h"
 #include <gnuradio/io_signature.h>
+#include <cmath>
 #include <cstdio>
 
 namespace gr {
@@ -48,12 +49,12 @@ tagged_decoder_impl::tagged_decoder_impl(generic_decoder::sptr my_decoder,
 
 int tagged_decoder_impl::calculate_output_stream_length(const gr_vector_int& ninput_items)
 {
-    if ((ninput_items[0] * d_decoder->rate()) > (d_mtu * 8)) {
+    if (std::lround(ninput_items[0] * d_decoder->rate()) > (d_mtu * 8)) {
         throw std::runtime_error("tagged_encoder: received frame is larger than MTU.");
     }
-    int diff =
-        d_decoder->rate() * d_decoder->get_input_size() - d_decoder->get_output_size();
-    d_decoder->set_frame_size(round(ninput_items[0] * d_decoder->rate()) - diff);
+    int diff = std::lround(d_decoder->rate() * d_decoder->get_input_size()) -
+               d_decoder->get_output_size();
+    d_decoder->set_frame_size(std::lround(ninput_items[0] * d_decoder->rate()) - diff);
     return d_decoder->get_output_size();
 }
 

